PMacc: refactor kernel en-queuing

psychocoderHPC opened the pull request:
Before this PR PMacc used cupla to execute the device kernel, this came with the drawback that it was not possible to execute an instance of a kernel. Cupla is default constructing the kernel even if PMacc was already using an instance for the kernel.
Therefore a kernel was not allowed to have data members.

This PR refactors the kernel execution method of PMacc and is forwarding the device kernel instance directly to alpaka.

ATTENTION: The interface is backward compatible and therefore PMacc and PIConGPU require no code changes.

It is now allowed to execute a device lambda as a kernel.

Example for a lambda hello world (one thread, one block):
```C++
PMACC_KERNEL([]ALPAKA_FN_ACC(auto const& acc) -> void{
    printf("Hello World.\n");
})(1,1)();
```

For example, delete all particles without cleaning the frame list.
```C++
constexpr uint32_t numWorkers
    = pmacc::traits::GetNumWorkers<pmacc::math::CT::volume<SuperCellSize>::type::value>::value;

auto const mapper = makeAreaMapper<CORE + BORDER>(this->cellDescription);

PMACC_KERNEL(
    [numWorkers] ALPAKA_FN_ACC(auto const& acc, auto particleBox, auto const mapper) -> void
    {
        uint32_t const workerIdx = cupla::threadIdx(acc).x;

        DataSpace<simDim> const superCellIdx(
            mapper.getSuperCellIndex(DataSpace<simDim>(cupla::blockIdx(acc))));

        // relative offset (in cells) to the supercell (including the guard)
        DataSpace<simDim> const superCellOffset = superCellIdx * SuperCellSize::toRT();

        auto forEachParticle
            = pmacc::particles::algorithm::acc::makeForEach<numWorkers>(workerIdx, particleBox, superCellIdx);

        forEachParticle(acc, [](auto const& accelerator, auto& particle) { particle[multiMask_] = 0; });
    })
(mapper.getGridDim(), numWorkers)(this->getDeviceParticlesBox(),mapper);
```

jkelling commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -44,124 +44,175 @@
+            }
+        };
+
+        //! Mata data of a device kernel
```
```suggestion
        //! Meta data of a device kernel
```

jkelling commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -44,124 +44,175 @@
         struct KernelStarter;
 
-        /** wrapper for the user kernel functor
+        /** Wrapper for the user kernel functor.
```
This not a sentence: the verb is missing.
```suggestion
        /** wrapper for the user kernel functor
```

jkelling commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -44,124 +44,175 @@
             }
 
-            /** configured kernel object
+            /** Configured kernel object.
```
not s sentence
```suggestion
            /** configured kernel object
```

jkelling commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -44,124 +44,175 @@
+            /** Configured kernel object.
              *
-             * this objects contains the functor and the starting parameter
+             * This objects contains the functor and the starting parameter.
```
```suggestion
             * This object contains the functor and the starting parameter.
```

jkelling commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -44,124 +44,175 @@
+
+        /** Configured kernel object.
+         *
+         * This objects contains the kernel functor, kernel meta information and the starting parameters.
```
Why the term "starting parameters"? I grant, that this was use before this PR, but "launch parameters" would sound much mor familiar.

psychocoderHPC replied:
bad eng. from my side :-)

psychocoderHPC commented:
This PR is reducing the cmem usage per kernel by 8 to 16 byte.

jkelling commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -44,124 +44,175 @@
+        /** Wrapper for the user kernel functor.
          *
-         * contains debug information like filename and line of the kernel call
+         * This objects contains the kernel functor, kernel meta information and the starting parameters.
```
This object does *not contain* kernel starting params, these are only passed to its `operator()` and incorporated into `KernelStarter`.

psychocoderHPC replied:
removed, left over from the old structure.

jkelling commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -44,124 +44,175 @@
              * @param gridExtent grid extent configuration for the kernel
              * @param blockExtent block extent configuration for the kernel
+             *
+             * @return object to bind arguments to a kernel
```
The returned objects does not have arguments to the kernel bound. It has only starting parameters bound. Its purpose is to offer functionality to bind arguments, but this detail should be part of that type's doc. Also, according to `KernelStarter`'s doc, it does also not bin arguments as far as the caller ID concerned. It just calls th kernel with provided arguments.

jkelling reviewed (changes requested):
Looks good.

I am not familiar with the `pmacc::TaskKernel` from which the kernel-invoking `operator()` gets a CUDA stream, which a cupla manager maps to an `alpaka::queue`, in which the `kernelTask`, which is an `alpaka::TaskKernel`, is enqueued. This seems very convoluted. Also what does `taskKernel->getCudaStream()` if the backend us not CUDA, e.g. Sycl?

Changes are only natural language fixed in comments and some semantical issues in some documentation comments.

On the topic of natural language: In the title of this PR `s/en-queue/enqueue/`.

steindev commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -44,124 +44,175 @@
                 pmacc::TaskKernel* taskKernel = pmacc::Environment<>::get().Factory().createTaskKernel(kernelName);
 
-                DataSpace<traits::GetNComponents<T_VectorGrid>::value> gridExtent(m_gridExtent);
+                auto gridExtent = cupla::IdxVec3(m_gridExtent.z, m_gridExtent.y, m_gridExtent.x);
+                auto blockExtent = cupla::IdxVec3(m_blockExtent.z, m_blockExtent.y, m_blockExtent.x);
```
Please add comment on the sense of this.

steindev commented on include/pmacc/eventSystem/events/kernelEvents.hpp:
```diff
@@ -173,39 +224,57 @@ namespace pmacc
+{
+    namespace trait
+    {
+        //! KernelWithDynSharedMem has defined the extern shared memory as member
```
Please clarify that this specializes a struct defined in alpaka.

steindev reviewed (changes requested):
Only little changes for clarification required.

psychocoderHPC commented:
> Also what does taskKernel->getCudaStream() if the backend us not CUDA, e.g. Sycl?

@jkelling The name is an old leftover from the past and is still not refactored. This PMacc part is giving a cupla stream depending on the internal event system state. It is valid for Sycl and CPU too.
unfortunately we have still many left over names from the last decade where we still name thinks CUDA but they are generic cupla/alpaka device parts.

> ...  in which the kernelTask, which is an alpaka::TaskKernel, is enqueued. This seems very convoluted

The complex structure is required because the event system is selecting the stream depending on the dependencies of tasks.

psychocoderHPC commented:
@steindev @jkelling Thanks for the review comments. I overworked this PR. I amended my changes because the PR is very small.

psychocoderHPC committed:
PMacc: refactor kernel en-queuing

Before this PR PMacc used cupla to execute the device kernel, this came with the drawback that it was not possible to execute an instance of a kernel. Cupla is default constructing the kernel even if PMacc was already using an instance for the kernel.
Therefore a kernel was not allowed to have data members.

This PR refactore the kernel execution method of PMacc and is forwarding the device kernel instance directly to alpaka.

ATTENTION: The interface is backward compatible and therefore PMacc and PIConGPU require no code changes.

It is now allowed to execute a device lambda as a kernel.
